src/query/storages/fuse/fuse/src/table_functions/fuse_snapshots/fuse_snapshot.rs

```diff
@@ -45,16 +49,45 @@ impl<'a> FuseSnapshot<'a> {
                 self.table.operator.clone(),
                 snapshot_version,
             );
-            let (snapshots, _) = snapshots_io
-                .read_snapshot_lites(
-                    snapshot_location,
-                    limit,
-                    false,
-                    snapshot.and_then(|s| s.timestamp),
-                    &|_| {},
-                )
-                .await?;
-            return self.to_block(&meta_location_generator, &snapshots, snapshot_version);
+            let snapshot_lite = match limit {
+                None => {
+                    // Use SnapshotIO only if limitation is None.
+                    //
+                    // SnapshotsIO lists snapshots from object storage, if we limit the number of
+                    // items being list , there might be the case that the snapshots returned
+                    // can not be chained together.
+                    let (snapshots, _) = snapshots_io
+                        .read_snapshot_lites(
+                            snapshot_location,
+                            None,
+                            false,
+                            snapshot.and_then(|s| s.timestamp),
+                            &|_| {},
+                        )
+                        .await?;
+                    Ok(snapshots)
+                }
+                Some(l) => {
+                    // SnapshotHistoryReader (which TableSnapshotReader impls) traverses the history
+                    // of snapshot sequentially, by using the TableSnapshot::prev_snapshot_id, which
+                    // guarantees the snapshot returned can be chained together
+                    let table_snapshot_reader = MetaReaders::table_snapshot_reader(
+                        self.ctx.get_data_operator()?.operator(),
+                    );
+                    table_snapshot_reader
```

Comment:
It seems read all snapshot first one by one and then take the limit? If so, ten thousand snapshots will take a few minutes if limit 10.

Reply:
not sure about this.  
usually, fetching tens of small files from object storage sequentially takes hundreds of ms. if the value of the limitation is small, maybe traversing them by `prev_snapshot_id` is better?

Reply:
oh... I misinterpreted your comment.

> If so, ten thousand snapshots will take a few minutes with limit 10.

no, it will not. in this case, only 10 snapshots will be traversed, from the latest one to the 10th. SnapshotHistoryReader returns a stream(which is lazy), and stream::take(10) only takes the first 10 elements from the stream.